libretroshare/src/services/p3discovery2.cc

```diff
@@ -1270,7 +1238,72 @@ bool p3discovery2::setPeerVersion(const SSLID &peerId, const std::string &versio
 	it->second.mVersion = version;
 	return true;
 }
-						  
+
+void p3discovery2::recvInvite(
+        std::unique_ptr<RsGossipDiscoveryInviteItem> inviteItem )
+{
+	RsPeerDetails dt; uint32_t eCode;
+	if( rsPeers
+	        && rsPeers->loadDetailsFromStringCert(
+	            inviteItem->mInvite, dt, eCode )
+	        && rsPeers->isFriendPendingPgp(dt.id) )
+	{ // notify we got a pendig PGP
+		typedef RsGossipDiscoveryPendingPgpFriendInviteReceivedEvent Evt_t;
+
+		// Ensure rsEvents is not deleted while we use it
+		std::shared_ptr<RsEvents> lockedRsEvents = rsEvents;
+		if(lockedRsEvents)
+			lockedRsEvents->postEvent(
+			            std::unique_ptr<Evt_t>(new Evt_t(inviteItem->mInvite)) );
+	}
+}
+
+void p3discovery2::rsEventsHandler(const RsEvent& event)
+{
+	switch(event.mType)
+	{
+	case RsEventType::AUTHSSL_CONNECTION_AUTENTICATION:
+	{
+		using Evt_t = RsAuthSslConnectionAutenticationEvent;
+		const Evt_t& evt = static_cast<const Evt_t&>(event);
+		if(evt.mIsPendingPpg)
+		{
+			Dbg1() << __PRETTY_FUNCTION__
+			       << " AUTHSSL_CONNECTION_AUTENTICATION event: " << event
+			       << std::endl;
+
+			const RsPeerId& sslId = evt.mSslId;
+			RsThread::async([sslId]()
+			{ /* Make sure the connection is ready before requesting invite. */
+				auto timeout = std::chrono::system_clock::now() +
+				        std::chrono::seconds(15);
+				while ( rsDisc && rsPeers &&
+				        std::chrono::system_clock::now() < timeout )
```

Comment:
This is not robust. You should use a queue and tick regularly to send the invite requests. You cannot ensure that your timeout will not be reached because of all sorts of delays that happen. So what you're doing is creatinga thread that will wait for a time that is either too short, or too long making potentially many threads to wait for the right conditions to continue, and eating file descriptors.

Reply:
Once a peer is authenticated the event is emitted, and the code there actually listen for that event, what happens is that even when rsPeers->isOnline(...) returns true the send keep failing for a bit more, and there is no event that tells when it become possible to send ATM and I could not find (damn unreadable code) at exactly what point it become possible to send an item to a peer without failing 100% of the times. Those sleeps are ugly to me too but I have added them empirically because I found no way to determine when it is possible to send an item to a peer (which apparently is a bit after RsPeer start reporting it as online -_- ). Also only one thread is created, so no problem with pid getting exhausted ;)

Reply:
No really, that is terrible. You see why right? The proper way to handle this here is to post the events into a queue and parse this queue in the data_tick() or tick() function of the service.

Reply:
I agree that the sleep thing is ugly, but putting it in a queue, doesn't help with the problem that we don't really know when the peer become available in terms of `sendItem`. RsPeers::isOnline(...) start returning true a little bit before `sendItem` would success... Where is the place where we know for sure it is possible to send an item to a peer? In that pace I could emit an event, and listen for that, without any sleep

Reply:
In tick() (or data_tick()?) you can query rsPeers->isPeerOnline() and when it says it is online, you can sent items without problem. We've been doing this for a while. If it was not working we would have noticed already.
Also, most of the time, when handling this kind of async queued events, we call the function that does this only once in a while (e.g. every 5, 10 or even 60 secs depending on the situation) and it's far likely that the peer is connected since several secs

Reply:
> you can query rsPeers->isPeerOnline() and when it says it is online, you can sent items without problem

No it is not like that, here I had to add the `sleep_for`
```
if(rsPeers->isOnline(sslId))
					{
						// Need to wait a bit more even if online
						std::this_thread::sleep_for(std::chrono::seconds(2));
						if(rsDisc && rsDisc->requestInvite(sslId, sslId))
``` 
becase the `sendItem(item)` inside `requestInvite` was failing (i see that becaus eit print an helpful message when that happens)

So doing this inside `tick()` maybe seems more elegant, but would work by chance anyway, most probably because tick is called once in a while... still we have the problem that `rsPeers->isOnline(sslId)` returns true a bit before we are able to send items

Reply:
the same test isPeerOnline() is used *everywhere* and we have never watched items not being sent because the peer is actually not online. What may have happenned in your case is that the item is of too low priority and therefore got stuck for a while before is was actually sent.
If you want to make sure to wait for some time after the peer is online, then add a field to your queue item with the last time you checked that it is online. When this time is long enough, send your item. But I dont even think that is necessary at all.
For the record the async[] trick is nice and works best in GUI situations especially because of Qt thread issues, but here it's really not what you need.